### PR TITLE
Sector Exposure

### PR DESCRIPTION
**Pull request by ACampbell22**
This adds the ability to sum positions by sector given a dictionary or series of symbol to sector mappings.

### CODE REVIEW

File: pyfolio/pos.py [hunk 1]
```diff
@@ -171,3 +173,35 @@ def get_turnover(transactions, positions, period=None):
...
+
+    Returns
+    -------
+    positions_alloc : pd.DataFrame
+        Sectors and their allocations.
+    """
+    unmapped_pos = np.setdiff1d(positions.drop('cash', axis=1).columns.values,
+                                symbol_sector_map.keys())
+    if len(unmapped_pos) > 0:
+        warn_message = "Warning: Symbols {} have no sector mapping.".format(
```
Comment (twiecki) [inline pyfolio/pos.py]: Maybe add that they will be ignored for the plot below.

File: pyfolio/tears.py [hunk 2]
```diff
@@ -328,16 +330,20 @@ def create_position_tear_sheet(returns, positions, gross_lev=None,
         If True, returns the figure that was plotted on.
     set_context : boolean, optional
         If True, set default plotting style context.
+    sector_mapping:
+        Dictionary or Series.
```
Comment (twiecki) [inline pyfolio/tears.py]: The type info should be moved up (see the others) and renamed to "dict or pandas.Series". Then the second line describes the data structure expected and what it results in.

File: pyfolio/tests/test_pos.py [hunk 3]
```diff
@@ -121,3 +125,41 @@ def test_get_turnover(self):
...
+
+    def test_sector_exposure(self):
+        """
+        Tests sector exposure mapping and rollup.
+
+        """
+        dates = date_range(start='2015-01-01', freq='D', periods=20)
+        positions = DataFrame([[1.0, 2.0, 3.0, 10.0]]*len(dates),
+                              columns=[0, 1, 2, 'cash'], index=dates)
+        mapping_d = {
```
Comment (twiecki) [inline pyfolio/tests/test_pos.py]: These test cases can be cleaner done with nose.parameterize (see some of the other tests for examples).

**Comment by twiecki** [comment]
Looks great, these are just some high-level comments. I'll do a more in-depth review soon.

File: pyfolio/plotting.py [hunk 4]
```diff
@@ -927,6 +927,49 @@ def show_and_plot_top_positions(returns, positions_alloc,
...
+    """Plots the sector exposures of the portfolio over time.
+
+    Parameters
+    ----------
+    returns : pd.Series
+        Daily returns of the strategy, noncumulative.
+         - See full explanation in tears.create_full_tear_sheet.
+    sector_alloc : pd.DataFrame
+        Portfolio allocation of positions. See pos.get_sector_alloc.
+    show_and_plot : int, optional
```
Comment (ssanderson) [inline pyfolio/plotting.py]: This kwarg appears to be missing?

File: pyfolio/pos.py [hunk 5]
```diff
@@ -171,3 +173,35 @@ def get_turnover(transactions, positions, period=None):
...
+    symbol_sector_map : dict
+        Sector to security identifier mapping.
+        Sectors as keys, security ids as values.
+
+    Returns
+    -------
+    positions_alloc : pd.DataFrame
+        Sectors and their allocations.
+    """
+    unmapped_pos = np.setdiff1d(positions.drop('cash', axis=1).columns.values,
```
Comment (ssanderson) [inline pyfolio/pos.py]: You drop `cash` twice in this function.  You probably want to just do that once and then save the result.

File: pyfolio/pos.py [hunk 6]
```diff
@@ -171,3 +173,35 @@ def get_turnover(transactions, positions, period=None):
...
+    """
+    Sum position exposures by sector.
+
+    Parameters
+    ----------
+    positions : pd.DataFrame
+        Contains position values or amounts.
+    symbol_sector_map : dict
+        Sector to security identifier mapping.
+        Sectors as keys, security ids as values.
```
Comment (ssanderson) [inline pyfolio/pos.py]: I don't think this description is correct if the intent here is to be summing by sector.  The values of your grouping dict should be your sectors, not the keys:

```
In [8]: df = DataFrame(arange(10).reshape(5, 2))

In [9]: df
Out[9]:
   0  1
0  0  1
1  2  3
2  4  5
3  6  7
4  8  9

In [10]: df = DataFrame(arange(10).reshape(5, 2), columns=['x', 'y'], index=['a', 'b', 'c', 'd', 'e'])

In [11]: df
Out[11]:
   x  y
a  0  1
b  2  3
c  4  5
d  6  7
e  8  9

In [12]: grouper = {'a': 'group_0', 'b': 'group_1', 'c': 'group_0', 'd': 'group_1', 'e': 'group_2'}

In [13]: df.groupby(by=grouper).sum()
Out[13]:
         x   y
group_0  4   6
group_1  8  10
group_2  8   9
```

File: pyfolio/tears.py [hunk 7]
```diff
@@ -352,6 +358,18 @@ def create_position_tear_sheet(returns, positions, gross_lev=None,
 
     plotting.plot_holdings(returns, positions_alloc, ax=ax_holdings)
 
+    if sector_mappings is not None:
+        sector_exposures = pos.get_sector_exposures(positions, sector_mappings)
+
+        if sector_exposures is None:
```
Comment (ssanderson) [inline pyfolio/tears.py]: I don't see a case where `sector_exposures` could return `None`.  What's the intent behind this check?

**Comment by twiecki** [comment]
I rebased the commit. Make sure to pull -f on this branch before making any changes.

File: pyfolio/plotting.py [hunk 8]
```diff
@@ -921,6 +921,46 @@ def show_and_plot_top_positions(returns, positions_alloc,
...
+    # Put a legend below current axis
+    ax.legend(
+        loc='upper center', frameon=True, bbox_to_anchor=(
+            0.5, -0.14), ncol=5)
+
+    df_cum_rets = timeseries.cum_returns(returns, starting_value=1)
+    ax.set_xlim((df_cum_rets.index[0], df_cum_rets.index[-1]))
+    ax.set_ylabel('Exposure by sector')
+
+
```
Comment (twiecki) [inline pyfolio/plotting.py]: It's not returning `ax`.

File: pyfolio/plotting.py [hunk 9]
```diff
@@ -921,6 +921,46 @@ def show_and_plot_top_positions(returns, positions_alloc,
...
+    sector_alloc : pd.DataFrame
+        Portfolio allocation of positions. See pos.get_sector_alloc.
+    ax : matplotlib.Axes, optional
+        Axes upon which to plot.
+    **kwargs, optional
+        Passed to plotting function.
+
+    Returns
+    -------
+    ax : matplotlib.Axes, conditional
```
Comment (twiecki) [inline pyfolio/plotting.py]: conditional?

File: pyfolio/plotting.py [hunk 10]
```diff
@@ -921,6 +921,46 @@ def show_and_plot_top_positions(returns, positions_alloc,
...
+    box = ax.get_position()
+    ax.set_position([box.x0, box.y0 + box.height * 0.1,
+                     box.width, box.height * 0.9])
+
+    # Put a legend below current axis
+    ax.legend(
+        loc='upper center', frameon=True, bbox_to_anchor=(
+            0.5, -0.14), ncol=5)
+
+    df_cum_rets = timeseries.cum_returns(returns, starting_value=1)
```
Comment (twiecki) [inline pyfolio/plotting.py]: Just for getting the index we don't need to convert to cum_rets.

File: pyfolio/pos.py [hunk 11]
```diff
@@ -163,3 +165,38 @@ def get_turnover(transactions, positions, period=None):
...
+        Contains position values or amounts.
+    symbol_sector_map : dict or pd.Series
+        Security identifier to sector mapping.
+        Security ids as keys/index, sectors as values.
+
+    Returns
+    -------
+    positions_alloc : pd.DataFrame
+        Sectors and their allocations.
+    """
```
Comment (twiecki) [inline pyfolio/pos.py]: Adding a small example self-contained example would be helpful here I think.

File: pyfolio/pos.py [hunk 12]
```diff
@@ -163,3 +165,38 @@ def get_turnover(transactions, positions, period=None):
...
+    cash = positions.pop('cash')
+
+    unmapped_pos = np.setdiff1d(positions.columns.values,
+                                symbol_sector_map.keys())
+    if len(unmapped_pos) > 0:
+        warn_message = """Warning: Symbols {} have no sector mapping.
+        They will not be included in sector allocations""".format(
+            " ".join(map(str, unmapped_pos)))
```
Comment (twiecki) [inline pyfolio/pos.py]: `", "`?

File: pyfolio/tears.py [hunk 13]
```diff
@@ -328,16 +330,21 @@ def create_position_tear_sheet(returns, positions, gross_lev=None,
         If True, returns the figure that was plotted on.
     set_context : boolean, optional
         If True, set default plotting style context.
+    sector_mapping: dict or pd.Series.
```
Comment (twiecki) [inline pyfolio/tears.py]: , optional

File: pyfolio/tests/test_pos.py [hunk 14]
```diff
@@ -121,3 +125,41 @@ def test_get_turnover(self):
         result = get_turnover(transactions, positions, period='M')
         expected = Series([10.0], index=index)
         assert_series_equal(result, expected)
+
+    @parameterized.expand([
```
Comment (twiecki) [inline pyfolio/tests/test_pos.py]: Nice!

**Comment by twiecki** [comment]
This looks great, see minor last comments and make sure to force-pull before updating. Thanks @ssanderson for helping with the review.

**Comment by twiecki** [comment]
Merged with https://github.com/quantopian/pyfolio/commit/a89178e87bdf75442a444930be9a143ccb4788fe. Thanks @ACampbell22.

**Pull request closed without merging**